src/Upstream.CommandLine/CommandLineApplication.cs

```diff
@@ -166,7 +168,19 @@ public CommandLineApplication UseExceptionHandler(Func<Exception, InvocationCont
         /// </remarks>
         internal Parser Build()
         {
-            return _builder.Build();
+            return _builder.Build(_serviceProviderFactory);
+        }
+
+        private static IServiceProvider MakeDefaultServiceProvider(IServiceCollection services)
+        {
+            return services.BuildServiceProvider();
+        }
+
+        public CommandLineApplication WithServiceProvider(
```

Comment:
I would rename this to `UseServiceProviderFactory` to better resemble the [HostBuilder method](https://learn.microsoft.com/en-us/dotnet/api/microsoft.extensions.hosting.hostbuilder.useserviceproviderfactory?view=dotnet-plat-ext-7.0). It would also be cool if we refactored this a little bit to support overloads with the [IServiceProviderFactory](https://learn.microsoft.com/en-us/dotnet/api/microsoft.extensions.dependencyinjection.iserviceproviderfactory-1?view=dotnet-plat-ext-7.0):
```csharp
CommandLineApplication WithServiceProvider(Func<IServiceCollection, IServiceProvider> providerFactory);

CommandLineApplication UseServiceProviderFactory<TContainerBuilder>(IServiceProviderFactory<TContainerBuilder> factory);

CommandLineApplication UseServiceProviderFactory<TContainerBuilder>(Func<HostBuilderContext, IServiceProviderFactory<TContainerBuilder>> factory);
```

Reply:
Also, it would be great if we added some xml doc comments as well:
```
/// <summary>
/// Overrides the factory used to create the service provider.
/// </summary>
```

Reply:
Regarding the overloads, I don't personally feel comfortable adding functionality I won't use. I can't get a good feel for the usability of it in that case. I also have limited time as well (new overloads means more unit tests). While I think your idea is good, I think I will pass on the additional overloads. Fortunately I think that would be a fantastic addition later on down the road by someone more familiar with them.

Regarding all of your other comments: Full agreement and I will make those changes. Thank you!

Reply:
So I'm digging more into this and I see an issue. I can't accept a `IServiceProviderFactory` because it's generic and I have to store it in the command line applicationn class. Which forces it to be generic as well.

I think a simpler solution would be to add a method to customize the host builder. I'll write something up and push it so you can review. But basically the code smell I see is that we're having to forward a lot of IHostBuilder methods through your classes which shouldn't be needed and it adds complexity. I'll see if I can simplify it.

Reply:
Ok I have pushed up a radically different solution. Please let me know if you think this is acceptable. If so, I will go back and add the necessary XML documentation. I did not do that initially in the interest of avoiding potential wasted effort.